File: python/1599.py
from typing import List


class Solution:
    def minOperationsMaxProfit(self, customers: List[int], boardingCost: int, runningCost: int) -> int:
        wait = 0
        earn = 0
        maxEarn = earn
        ans = -1
        i = 0
        while i < len(customers) or wait > 0:
            if i < len(customers):
                wait += customers[i]
            passanger = min(wait, 4)
            wait -= passanger
            earn += passanger * boardingCost - runningCost
            if earn > maxEarn:
                earn = maxEarn
                ans = i
            i+=1
        return ans+1 if ans != -1 else -1

    
s = Solution()
customers = [8,3,8]
boardingCost = 5
runningCost = 6
print(s.minOperationsMaxProfit(customers, boardingCost, runningCost))

File: go/1599.go
package main

func minOperationsMaxProfit(customers []int, boardingCost int, runningCost int) int {
	if boardingCost*4 < runningCost {
		return -1
	}
	remain := 0
	for i := range customers {
		if customers[i] >= 4 {
			remain += customers[i] - 4
			customers[i] = 4
		} else {
			if customers[i]+remain >= 4 {
				remain -= 4 - customers[i]
				customers[i] = 4
			} else {
				customers[i] += remain
				remain = 0
			}
		}
	}
	for remain > 0 {
		if remain >= 4 {
			customers = append(customers, 4)
			remain -= 4
		} else {
			customers = append(customers, remain)
			remain = 0
		}
	}
	ans := -1
	ansT := 0
	tmp := 0
	for i, x := range customers {
		tmp += boardingCost*x - runningCost
		if tmp > 0 && tmp > ansT {
			ansT = tmp
			ans = i
		}
	}
	return ans
}
